promql/quantile.go

```diff
@@ -158,9 +158,18 @@ func histogramQuantile(q float64, h *histogram.FloatHistogram) float64 {
 	var (
 		bucket histogram.Bucket[float64]
 		count  float64
-		it     = h.AllBucketIterator()
-		rank   = q * h.Count
+		it     histogram.BucketIterator[float64]
+		rank   float64
 	)
+
+	if q < 0.5 {
```

Comment:
We also have to use the forward iterator if there are NaN observations in the histogram (because they are not included in the h.Count). You can detect that case by checking if h.Sum is NaN.